xmview-os/um_cmd/viewsu.1
.\" Copyright (c) 2009 Renzo Davoli
.\"
.\" This is free documentation; you can redistribute it and/or
.\" modify it under the terms of the GNU General Public License,
.\" version 2, as published by the Free Software Foundation.
.\"
.\" The GNU General Public License's references to "object code"
.\" and "executables" are to be interpreted as the output of any
.\" document formatting or typesetting system, including
.\" intermediate and printed output.
.\"
.\" This manual is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public
.\" License along with this manual; if not, write to the Free
.\" Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston,
.\" MA 02110-1301 USA.

.TH VIEW-OS 1 "August 5, 2009" "VIEW-OS: a process with a view"
.SH NAME
viewsu \- change (virtual) user ID or become (virtual) superuser
.SH "SYNOPSIS"
.HP \w'\fBviewsu\fR\ 'u
\fBviewsu\fR [\fIoptions\fR] [\fIusername\fR]
.SH "DESCRIPTION"
.PP
The
\fBviewsu\fR
command is used to become another user during a login session\&. Invoked without a
\fBusername\fR,
\fBviewsu\fR
defaults to becoming the superuser\&. The optional argument
\fB\-\fR
may be used to provide an environment similar to what the user would expect had the user logged in directly\&.
This command is the View-OS counterpart of the su(1) command.
The value of
\fB$PATH\fR
is reset to
/bin:/usr/bin
for normal users, or
/sbin:/bin:/usr/sbin:/usr/bin
for the superuser\&. This may be changed with the
\fIENV_PATH\fR
and
\fIENV_SUPATH\fR
definitions in
/etc/login\&.defs\&.
.SH OPTIONS
.PP
The options which apply to the
\fBviewsu\fR
command are:
.PP
\fB\-c\fR, \fB\-\-command\fR \fICOMMAND\fR
.RS 4
Specify a command that will be invoked by the shell using its
\fB\-c\fR\&.
.RE
.PP
\fB\-\fR, \fB\-l\fR, \fB\-\-login\fR
.RS 4
Provide an environment similar to what the user would expect had the user logged in directly\&.
.sp
When
\fB\-\fR
is used, it must be specified as the last
\fBviewsu\fR
option\&. The other forms (\fB\-l\fR
and
\fB\-\-login\fR) do not have this restriction\&.
.RE
.PP
\fB\-s\fR, \fB\-\-shell\fR \fISHELL\fR
.RS 4
The shell that will be invoked\&.
.sp
The invoked shell is chosen from (highest priority first):
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The shell specified with \-\-shell\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.\" .el \{\
.\" .sp -1
.\" .IP \(bu 2.3
.\" .\}
.\" If
.\" \fB\-\-preserve\-environment\fR
.\" is used, the shell specified by the
.\" \fB$SHELL\fR
.\" environment variable\&.
.\" .RE
.\" .sp
.\" .RS 4
.\" .ie n \{\
.\" \h'-04'\(bu\h'+03'\c
.\" .\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}
The shell indicated in the
/etc/passwd
entry for the target user\&.
.RE
.sp
.RS 4
.ie n \{\
\h'-04'\(bu\h'+03'\c
.\}
.el \{\
.sp -1
.IP \(bu 2.3
.\}

/bin/sh
if a shell could not be found by any above method\&.
.RE
.RS 4

.SH SEE ALSO
.BR su(1),
.BR linux.defs(5)
.SH AUTHORS
View-OS is a project of the Computer Science Department, University of
Bologna. Project Leader: Renzo Davoli.
.br
<http://www.sourceforge.net/projects/view-os>

Howto's and further information can be found on the project wiki
<wiki.virtualsquare.org>.
.SH NOTE
Most part of the text is taken from su(1).





0.1/NEWS
Bug fixed in fat_write_data
Bug fixed in fat_mkdir (zeroing of new allocated clusters)

xmview-os/um_cmd/viewsu.c
#include<stdio.h>
#include<stdlib.h>
#include<unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <getopt.h>
#include <pwd.h>
#include <string.h>
#include <libgen.h>
#include <um_lib.h>

static struct option long_options[] = {
	{"command", 1, 0, 'c'},
	{"login", 0, 0, 'l'},
	{"shell", 1, 0, 's'},
	{"preserve-environment", 0, 0, 'p'},
	{"help", 1, 0, 'h'},
	{0, 0, 0, 0}
};

char *command;
int login;
char *shell;
int preserve_environment;
char *user;
int uid;
int gid;
char *arg0;

void usage(char *argv0)
{
	char *name=basename(argv0);
	fprintf(stderr,
			"Usage: %s [options] [LOGIN]\n"
			"\n"
			"Options:\n"
			"  -c, --command COMMAND         pass COMMAND to the invoked shell\n"
			"  -h, --help                    display this help message and exit\n"
			"  -, -l, --login                make the shell a login shell\n"
			//"  -m, -p,\n"
			//"  --preserve-environment        do not reset environment variables, and\n"
			"                                keep the same shell\n"
			"  -s, --shell SHELL             use SHELL instead of the default in passwd\n"
			"\n",
			name
			);
	exit(2);
}

void setpath(void)
{
	char buf[1024];
	char *tag=(uid==0)?"ENV_SUPATH":"ENV_PATH";
	FILE *f=fopen("/etc/login.defs","r");
	char *path;
	if (f) {
		char *s;
		while((s=fgets(buf,1024,f)) != NULL) {
			while (*s==' ' || *s=='\t') s++;
			if (*s=='#') continue;
			if (strncmp(tag,s,strlen(tag))!=0) continue;
			s+=strlen(tag);
			if (*s != ' ' && *s != '\t') continue;
			while (*s==' ' || *s=='\t') s++;
			if (strncmp("PATH=",s,5)!=0) continue;
			s+=5;
			s[strlen(s)-1]=0;
			path=s;
			break;
		}
		fclose(f);
	}
	if (!path) {
		if (uid)
			path="/bin:/usr/bin";
		else
			path="/sbin:/bin:/usr/sbin:/usr/bin";
	}
	setenv("PATH",path,1);
}

int main(int argc, char *argv[])
{
	pid_t pid;
	int status;
	int c;
	struct passwd *pwd;
	
	/* outside viewos use su(1) */
	if (um_check_viewos()==0) 
		execvp("su",argv);
	
	while (1) {
		int option_index = 0;
		c = getopt_long(argc, argv, "c:ls:mph",
				long_options, &option_index);
		if (c == -1)
			break;
		switch (c) {
			case 'c': command=optarg;
								break;
			case 'l': login=1;
								break;
			case 's': shell=optarg;
								break;
		//	case 'm':
		//	case 'p': preserve_environment=1;
		//						break;
			case 'h': usage(argv[0]);
								break;
		}
	}

	if (argc > optind+1)
		usage(argv[0]);

	if (argc > optind) {
		if (strcmp(argv[optind],"-")==0)
			login=1;
		else
			user=argv[optind];
	}

	//if (!preserve_environment)
		//clearenv();
	if (user) {
		pwd=getpwnam(user);
		if (pwd == NULL) {
			fprintf(stderr,"Unknown id: %s\n",user);
			exit(1);
		}
		uid=pwd->pw_uid;
		gid=pwd->pw_gid;
		shell=pwd->pw_shell;
	} else {
		pwd=getpwuid(0);
		if(pwd)
			shell=pwd->pw_shell;
	}
	if (pwd)
		setenv("HOME",pwd->pw_dir,1);
	if (!shell) 
		shell="/bin/sh";
	setenv("SHELL",shell,1);
	if (login)
		asprintf(&arg0,"-%s",shell);
	else
		arg0=shell;

	switch (pid=fork()) {
		case -1: exit(1);
		case 0: 
						 setresuid(uid,uid,uid);
						 setresgid(gid,gid,gid);
						 setpath();
						 if (command)
							 execl(shell,arg0,"-c",command,0);
						 else
							 execl(shell,arg0,0);
						 break;
		default:
						 waitpid(pid,&status,0);
	}
}
